x/ibc/23-commitment/types/merkle.go

```diff
@@ -154,31 +183,292 @@ func (MerkleProof) GetCommitmentType() exported.Type {
+				if !ok {
+					return sdkerrors.Wrap(ErrInvalidProof, "proof is not an existence proof")
+				}
+				subroot, err = existProof.Exist.Calculate()
+				if err != nil {
+					return sdkerrors.Wrap(ErrInvalidProof, err.Error())
+				}
+
+				if !ics23.VerifyMembership(specs[i], subroot, p, subpath, value) {
+					return sdkerrors.Wrapf(ErrInvalidProof, "invalid proof for path: %s", path.String())
+				}
+			} else {
+				// The final proof in the chain will prove inclusion against the given root.
+				if !ics23.VerifyMembership(specs[i], root.GetHash(), p, subpath, value) {
+					return sdkerrors.Wrapf(ErrInvalidProof, "invalid proof for path: %s", path.String())
+				}
+			}
+		}
+		// Each subtree root becomes the proving value for the next proof in the chaining process
+		value = subroot
+	}
+
+	return nil
+
 }
 
-// IsEmpty returns true if the root is empty
+// BatchVerifyMembership verifies existence of a batch of items in a subtree for a given root with all items under the given path.
+// NOTE: All items must be part of a batch proof in the first chained proof, i.e. items must all be part of smallest subtree in the chained proof
+// NOTE: The path passed in must be the path from the root to the smallest subtree in the chained proof
+func (proof MerkleProof) BatchVerifyMembership(specs []*ics23.ProofSpec, root exported.Root, path exported.Path, items map[string][]byte) error {
+	if proof.IsEmpty() || root == nil || root.IsEmpty() || path == nil || path.IsEmpty() {
+		return sdkerrors.Wrap(ErrInvalidProof, "empty params or proof")
+	}
+
+	mpath, ok := path.(MerklePath)
+	if !ok {
+		return sdkerrors.Wrap(ErrInvalidProof, "path is not a merkle path for a merkle proof")
+	}
+
+	// Path length must be 1 less than proof length since paths for last proof are provided by item keys
+	if len(proof.Proofs) != len(mpath.KeyPaths)-1 || len(proof.Proofs) != len(specs) {
+		return sdkerrors.Wrapf(ErrInvalidProof, "invalid chained proof. chained proof length %d, spec length %d, path length %d must all be equal",
+			len(proof.Proofs), len(specs), len(mpath.KeyPaths)+1)
+	}
+
+	var value, subroot []byte
+	var err error
+	for i, p := range proof.Proofs {
+		// First proof in chain must be a batch proof containing existence proofs for all items
+		if i == 0 {
+			// For subtree verification, we simply calculate the root from the proof and use it to prove
+			// against the value
+			bproof, ok := p.Proof.(*ics23.CommitmentProof_Batch)
+			if ok {
+				if len(bproof.Batch.GetEntries()) == 0 || bproof.Batch.GetEntries()[0] == nil {
+					return sdkerrors.Wrap(ErrInvalidProof, "batch proof has empty entry")
+				}
+				bexist := bproof.Batch.GetEntries()[0].GetExist()
+				if bexist == nil {
+					return sdkerrors.Wrap(ErrInvalidProof, "batch proof is not an existence proof")
+				}
+				subroot, err = bexist.Calculate()
+			} else {
+				return sdkerrors.Wrap(ErrInvalidProof, "not a batch proof, compressed batch proof currently unimplemented")
```

Comment:
Not doing compressed batch proof calculation manually here but similar principle